File: solutions/elixir/allergies/1/lib/allergies.ex
defmodule Allergies do
  import Bitwise

  @allergens ~w(eggs peanuts shellfish strawberries tomatoes chocolate pollen cats)

  @doc """
  List the allergies for which the corresponding flag bit is true.
  """
  @spec list(non_neg_integer) :: [String.t()]
  def list(flags) do
    {_, allergies} =
      Enum.reduce(@allergens, {flags, []}, fn allergen, {n, acc} ->
        {bsr(n, 1), if(band(n, 1) == 1, do: [allergen | acc], else: acc)}
      end)

    allergies
  end

  @doc """
  Returns whether the corresponding flag bit in 'flags' is set for the item.
  """
  @spec allergic_to?(non_neg_integer, String.t()) :: boolean
  def allergic_to?(flags, item), do: flags |> list() |> Enum.member?(item)
end

File: solutions/erlang/allergies/2/allergies.erl
-module(allergies).
-export([allergies/1, isAllergicTo/2]).

-define(ALLERGIES, [eggs, peanuts, shellfish, strawberries, tomatoes, chocolate, pollen, cats]).

allergies(Flag) ->
  <<Cats:1, Pollen:1, Chocolate:1, Tomatoes:1, Strawberries:1, Shellfish:1, Peanuts:1, Eggs:1>> = <<Flag:8/integer>>,
  Flags = [Eggs, Peanuts, Shellfish, Strawberries, Tomatoes, Chocolate, Pollen, Cats],
  Allergens = lists:zipwith(fun parse/2, Flags, ?ALLERGIES),
  lists:filter(fun(A) -> A =/= not_allergic end, Allergens).

isAllergicTo(Allergen, Flag) ->
  lists:member(Allergen, allergies(Flag)).

parse(1, Allergen) -> Allergen;
parse(0, _Allergen) -> not_allergic.